ledger/acctupdates.go

```diff
@@ -1374,82 +1434,164 @@ func (au *accountUpdates) newBlockImpl(blk bookkeeping.Block, delta StateDelta)
+			for offset > 0 {
+				offset--
+				d, ok := au.deltas[offset][addr]
+				if ok {
+					return d.new, nil
+				}
 			}
 		}
-	}
 
-	// No updates of this account in the in-memory deltas; use on-disk DB.
-	// The check in roundOffset() made sure the round is exactly the one
-	// present in the on-disk DB.  As an optimization, we avoid creating
-	// a separate transaction here, and directly use a prepared SQL query
-	// against the database.
-	return au.accountsq.lookup(addr)
+		if syncronized {
+			au.accountsMu.RUnlock()
+			needUnlock = false
+		}
+		var dbRound basics.Round
+		// No updates of this account in the in-memory deltas; use on-disk DB.
+		// The check in roundOffset() made sure the round is exactly the one
+		// present in the on-disk DB.  As an optimization, we avoid creating
+		// a separate transaction here, and directly use a prepared SQL query
+		// against the database.
+		data, dbRound, err = au.accountsq.lookup(addr)
+		if dbRound == currentDbRound {
+			return data, err
+		}
+		if syncronized {
+			if dbRound < currentDbRound {
+				au.log.Errorf("accountUpdates.lookupImpl: database round %d is behind in-memory round %d", dbRound, currentDbRound)
+				return basics.AccountData{}, &StaleDatabaseRoundError{databaseRound: dbRound, memoryRound: currentDbRound}
+			}
+			au.accountsMu.RLock()
+			needUnlock = true
+			for currentDbRound >= au.dbRound && currentDeltaLen == len(au.deltas) {
+				au.accountsReadCond.Wait()
+			}
+		} else {
+			// in non-sync mode, we don't wait since we already assume that we're syncronized.
+			au.log.Errorf("accountUpdates.lookupImpl: database round %d mismatching in-memory round %d", dbRound, currentDbRound)
+			return basics.AccountData{}, &MismatchingDatabaseRoundError{databaseRound: dbRound, memoryRound: currentDbRound}
+		}
+	}
 }
 
 // getCreatorForRoundImpl returns the asset/app creator for a given asset/app index at a given round
-func (au *accountUpdates) getCreatorForRoundImpl(rnd basics.Round, cidx basics.CreatableIndex, ctype basics.CreatableType) (creator basics.Address, ok bool, err error) {
-	offset, err := au.roundOffset(rnd)
-	if err != nil {
-		return basics.Address{}, false, err
+func (au *accountUpdates) getCreatorForRoundImpl(rnd basics.Round, cidx basics.CreatableIndex, ctype basics.CreatableType, syncronized bool) (creator basics.Address, ok bool, err error) {
+	unlock := false
+	if syncronized {
+		au.accountsMu.RLock()
+		unlock = true
 	}
-
-	// If this is the most recent round, au.creatables has will have the latest
-	// state and we can skip scanning backwards over creatableDeltas
-	if offset == uint64(len(au.deltas)) {
-		// Check if we already have the asset/creator in cache
-		creatableDelta, ok := au.creatables[cidx]
-		if ok {
-			if creatableDelta.created && creatableDelta.ctype == ctype {
-				return creatableDelta.creator, true, nil
-			}
-			return basics.Address{}, false, nil
+	defer func() {
```

Comment:
nit: moving this `defer` inside the `if` cond and restoring the lock right after `au.accountsq.lookupCreator` will allow to get rid of `unlock` variable.

Reply:
I was considering that, but given that the loop could take multiple iterations, we don't want to defer the function call multiple times, right ?